rust/bin/agent/src/handler/upsert.rs

```diff
@@ -18,147 +18,179 @@ use log::{info, warn};
+            None => return Err(Status::invalid_argument("Missing vector in request")),
+        };
+        let s_inner = s.read().await;
+        let uuid = vec.id.clone();
+        if vec.vector.len() != s_inner.get_dimension_size() {
+            let err = Error::IncompatibleDimensionSize {
+                got: vec.vector.len(),
+                want: s_inner.get_dimension_size(),
+            };
+            let mut err_details = ErrorDetails::new();
+            let metadata = HashMap::new();
+            let resource_type = format!("{}/qbg.Upsert", resource_type);
+            let resource_name = format!("{}: {}({})", api_name, name, ip);
+            err_details.set_error_info(err.to_string(), domain, metadata);
+            err_details.set_request_info(
+                uuid,
+                String::from_utf8(request.encode_to_vec())
+                    .unwrap_or_else(|_| "<invalid UTF-8>".to_string()),
+            );
+            err_details.set_bad_request(vec![FieldViolation::new(
+                "vector dimension size",
+                err.to_string(),
+            )]);
+            err_details.set_resource_info(resource_type, resource_name, "", "");
+            let status = Status::with_error_details(
+                Code::InvalidArgument,
+                "Upsert API Incompatible Dimension Size detected",
+                err_details,
+            );
+            warn!("{:?}", status);
+            return Err(status);
+        }
+        if uuid.len() == 0 {
+            let err = Error::InvalidUUID { uuid: uuid.clone() };
+            let mut err_details = ErrorDetails::new();
+            let metadata = HashMap::new();
+            let resource_type = format!("{}/qbg.Upsert", resource_type);
+            let resource_name = format!("{}: {}({})", api_name, name, ip);
+            err_details.set_error_info(err.to_string(), domain, metadata);
+            err_details.set_request_info(
+                uuid.clone(),
+                String::from_utf8(request.encode_to_vec())
+                    .unwrap_or_else(|_| "<invalid UTF-8>".to_string()),
+            );
+            err_details.set_bad_request(vec![FieldViolation::new("uuid", err.to_string())]);
+            err_details.set_resource_info(resource_type, resource_name, "", "");
+            let status = Status::with_error_details(
+                Code::InvalidArgument,
+                format!("Upsert API invalid argument for uuid \"{}\" detected", uuid),
+                err_details,
+            );
+            warn!("{:?}", status);
+            return Err(status);
+        }
+        let rt_name;
+        let result;
+        let exists = s_inner.exists(uuid.clone());
+        if exists {
+            result = update_fn(
+                s.clone(),
+                resource_type,
+                api_name,
+                name,
+                ip,
+                &update::Request {
+                    vector: request.vector.clone(),
```

Comment:
we can use already cloned vec here
```suggestion
                    vector: vec.clone(),
```